### PR TITLE
Do not allow rebooting via mavlink when armed

### PR DESCRIPTION
**Pull request by peterbarker**
Should replace https://github.com/ArduPilot/ardupilot/pull/9941

### CODE REVIEW

**Comment by WickedShell** [comment]
Okay, now I'm going to hold to https://github.com/ArduPilot/ardupilot/pull/10633#issuecomment-467288136 and say we should just do this right...

File: Tools/autotest/common.py [hunk 1]
```diff
@@ -267,10 +267,35 @@ def fetch_parameters(self):
...
+                                       0,
+                                       0,
+                                       0,
+                                       0,
+                                       0)
+        self.detect_and_handle_reboot(old_bootcount)
+
     def reboot_sitl(self):
-        """Reboot SITL instance and wait it to reconnect."""
+        """Reboot SITL instance wait it to reconnect."""
```
Comment (OXINARF) [inline Tools/autotest/common.py]: Why the comment change? Looks worse.

File: Tools/autotest/common.py [hunk 2]
```diff
@@ -778,12 +803,16 @@ def arm_vehicle(self, timeout=20):
                 return True
         raise AutoTestTimeoutException("Failed to ARM with mavlink")
 
-    def disarm_vehicle(self, timeout=20):
+    def disarm_vehicle(self, timeout=20, force=False):
         """Disarm vehicle with mavlink disarm message."""
         self.progress("Disarm motors with MAVLink cmd")
+        p2 = 0
+        if force:
+            magic_force_disarm_value = 21196
```
Comment (OXINARF) [inline Tools/autotest/common.py]: Do we really need this temporary variable?

File: Tools/autotest/common.py [hunk 3]
```diff
@@ -778,12 +803,16 @@ def arm_vehicle(self, timeout=20):
...
+            p2 = magic_force_disarm_value
```
Comment (OXINARF) [inline Tools/autotest/common.py]: Just noting that this is the deprecated way of force disarm.

**Comment by peterbarker** [comment]
@IamPete1 Could you eyeball the Tracker changes in here, please?

**Comment by peterbarker** [comment]
> It looks generally OK. I'll say that the autotest changes seem unnecessarily complicated, I can't see any specific case (other than testing reboot) where we wouldn't want reboot to force the disarm.

Yeah, originally I did have `reboot_sitl` force the reboot if required.  I decided a forced-disarm where needed was OK, and that an implicit forced disarm wasn't in good taste.  Perhaps I was wrong there.  There are lots of places where we would want to know if the vehicle didn't reboot cleanly where we expected it to.

File: Tools/autotest/common.py (hunk 1 above)
Comment (peterbarker) [reply Tools/autotest/common.py]: Bad patch...  I've no idea what happened there, I *was* actually trying to fix that comment!

**Comment by IamPete1** [comment]
looks good to me, i have the same change to MAV_CMD_COMPONENT_ARM_DISARM in my PR, i not sure what MAV_COMP_ID_SYSTEM_CONTROL does, only that removing it allowed arming and disarming. I guess its deprecated?

File: Tools/autotest/common.py (hunk 2 above)
Comment (peterbarker) [reply Tools/autotest/common.py]: Will remove.

File: Tools/autotest/common.py (hunk 3 above)
Comment (peterbarker) [reply Tools/autotest/common.py]: Remind me which mechanism replaced it?

**Comment by peterbarker** [comment]
@IamPete1 I'm fighting the mavlink spec on component IDs ATM.  Essentially the spec says that if the FC hasn't seen any other component saying it handled SYSTEM_CONTROL then it should handle it itself.  Of course, the spec *also* says that a GCS shouldn't send targetted messages to a system/component it hasn't seen heartbeats from - and we don't heartbeat from anything other than our primary component ID.

I'm actually going to be adding a parameter to ardupilot which, when set, will indicate we *only* accept packets to our own component ID.

**Comment by OXINARF** [comment]
> There are lots of places where we would want to know if the vehicle didn't reboot cleanly where we expected it to.

@peterbarker Do we really care about that or do we care that the vehicle disarmed? It's too different things. Other than testing reboot itself, where do we care if the reboot is clean? If it is asking to reboot, I think we don't care if the vehicle is flying or not, the autotest just needs to reboot.

File: Tools/autotest/common.py (hunk 3 above)
Comment (OXINARF) [reply Tools/autotest/common.py]: MAV_CMD_DO_FLIGHTTERMINATION

**Comment by peterbarker** [comment]
On Thu, 28 Feb 2019, Francisco Ferreira wrote:

> @peterbarker Do we really care about that or do we care that the vehicle disarmed? It's too different things. Other than testing reboot itself, where do
> we care if the reboot is clean? If it is asking to reboot, I think we don't care if the vehicle is flying or not, the autotest just needs to reboot.

Well, that was the whole thing about the way I was treating reboot - a 
really nice and easy way (lazy?) way to end a test and move on.  Someone 
shifted the posts and decided mid-air reboots weren't a good idea....

Perhaps we shouldn't use "reboot" but "self.lazy_end_test()" or something 
- for when we really don't care.

**Comment by peterbarker** [comment]
On Thu, 28 Feb 2019, Francisco Ferreira wrote:

> +            magic_force_disarm_value = 21196
> +            p2 = magic_force_disarm_value
> 
> MAV_CMD_DO_FLIGHTTERMINATION

?  This magic value isn't that.

**Comment by OXINARF** [comment]
> Perhaps we shouldn't use "reboot" but "self.lazy_end_test()" or something - for when we really don't care.

Sounds like a good idea. What we want there is a clean SITL instance, killing it or rebooting doesn't really matter from the test point of view.

> ? This magic value isn't that.

It wasn't me writing it... https://github.com/ArduPilot/ardupilot/commit/a42674b0b05abe92486edb44ad16fd55d95c015e

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 4]
```diff
@@ -2504,6 +2504,11 @@ void GCS_MAVLINK::zero_rc_outputs()
  */
 MAV_RESULT GCS_MAVLINK::handle_preflight_reboot(const mavlink_command_long_t &packet)
 {
+    if (hal.util->get_soft_armed()) {
+        // refuse reboot when armed
+        return MAV_RESULT_FAILED;
```
Comment (magicrub) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: Does this also fail when in SITL?

**Comment by peterbarker** [comment]
On Sun, 3 Mar 2019, Tom Pittenger wrote:

>  MAV_RESULT GCS_MAVLINK::handle_preflight_reboot(const mavlink_command_long_
> t &packet)
>  {
> +    if (hal.util->get_soft_armed()) {
> +        // refuse reboot when armed
> +        return MAV_RESULT_FAILED;
> 
> Does this also fail when in SITL?

Yes.

**Pull request merged**